plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,517 @@
+    g_autoptr(GByteArray) fw_payload = g_byte_array_new();
+    gboolean result;
+
+    /* progress */
+    fu_progress_set_id(progress, G_STRLOC);
+    fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+    fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 15, NULL);
+    fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 95, NULL);
+
+    result = fu_algoltek_device_CMD03(self, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x09 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD04(self, 2, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x20_2 : ");
+        return FALSE;
+    }
+
+    fu_device_sleep(FU_DEVICE(self), 900);
+
+    result = fu_algoltek_device_CMD05(self, 0x80AD, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80AD : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80C0, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80C0 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80C9, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80C9 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80D1, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80D1 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80D9, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80D9 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80E1, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80E1 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD05(self, 0x80E9, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x07_80E9 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD04(self, 0, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x20_0 : ");
+        return FALSE;
+    }
+
+    fu_device_sleep(FU_DEVICE(self), 500);
+
+    /* get ISP image */
+    blob_ISP = fu_firmware_get_image_by_id_bytes(firmware, "ISP", error);
+    if (blob_ISP == NULL)
+        return FALSE;
+    fu_byte_array_append_bytes(fw_ISP, blob_ISP);
+    result = fu_algoltek_device_CMD06(self, fw_ISP, 0x2000, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x13 : ");
+        return FALSE;
+    }
+
+    result = fu_algoltek_device_CMD07(self, 0x2000, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x1D_2000 : ");
+        return FALSE;
+    }
+
+    fu_device_sleep(FU_DEVICE(self), 1000);
+
+    result = fu_algoltek_device_CMD08(self, error);
+    if (!result) {
+        g_prefix_error(error, "Error Code 0x19 : ");
+        return FALSE;
+    }
+
+    fu_device_sleep(FU_DEVICE(self), 500);
```

Comment:
You've got a lot of random sleeps between steps. Are these all mandatory? Is there any register you can poll for completion in between steps perhaps?

Also with the obfuscated cmd names I think it's a lot harder to follow. Could you please give them more descriptive names and the functions that use them adopt them as well?

Reply:
The firmware is programmed into the internal SPI flash and after each time laptop sends the data to ASIC, the ASIC needs time to do the data verification and writing of SPI flash.

Reply:
Is there some bit that the ASIC will set to say it's done with the data verification and write?  It would be better to poll that rather than a random wait if so.

Reply:
No, we didn't implemented this function, which is retrieving the writing status of SPI Flash, inside the ASIC circuit.